# Hypothesis Testing: Process We can Follow

So, Hypothesis testing is fundamental process in data sciencie for marking data-driven decisions and inferences about populations based on sample data. Below is the process we can follow for the task of Hypothesis Testing:
1. Gather the necessary data required for the hypothesis test
2. Define Null (H0) and Alternative Hypothesis (H1 or Ha)
3. Choose the Significance level (a), which is the propability of rejecting the null hypothesis when it is true
4. Select the appropiate statitical tests. Examples include t-tests for comparing means, chi-square tests for categorical data, and ANOVA for comparing means across more than two groups
5. Perform the chosen statistical test on your data
6. Determine the p-value and interpret the results of you statistical tests 

To get started with Hypothesis Testing, we need appropiate data. I found an ideal dataset for this task. You can download the dataset from here

## Hypothesis Testing using Python

Now, let's get started with the task of Hypothesis Testing by importing the necessary Python libraries and the dataset

In [4]:
import pandas as pd
from scipy.stats import ttest_ind

df = pd.read_csv("website_ab_test.csv")

print(df.head())

         Theme  Click Through Rate  Conversion Rate  Bounce Rate  \
0  Light Theme            0.054920         0.282367     0.405085   
1  Light Theme            0.113932         0.032973     0.732759   
2   Dark Theme            0.323352         0.178763     0.296543   
3  Light Theme            0.485836         0.325225     0.245001   
4  Light Theme            0.034783         0.196766     0.765100   

   Scroll_Depth  Age   Location  Session_Duration Purchases Added_to_Cart  
0     72.489458   25    Chennai              1535        No           Yes  
1     61.858568   19       Pune               303        No           Yes  
2     45.737376   47    Chennai               563       Yes           Yes  
3     76.305298   58       Pune               385       Yes            No  
4     48.927407   25  New Delhi              1437        No            No  


So, the dataset is based on the performance of two themens on a website. Our task is to find which theme performs better using Hypothesis Testing. Let's go trough the summary of the dataset, including the number of records, the presence of missing values, and basic statistics for the numerical columns

In [5]:
# dataset summary

summary = {
    'Number of Records': df.shape[0],
    'Number of Columns': df.shape[1],
    'Missing Values': df.isnull().sum(),
    'Numerical Columns Summary': df.describe()
}

summary

{'Number of Records': 1000,
 'Number of Columns': 10,
 'Missing Values': Theme                 0
 Click Through Rate    0
 Conversion Rate       0
 Bounce Rate           0
 Scroll_Depth          0
 Age                   0
 Location              0
 Session_Duration      0
 Purchases             0
 Added_to_Cart         0
 dtype: int64,
 'Numerical Columns Summary':        Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
 count         1000.000000      1000.000000  1000.000000   1000.000000   
 mean             0.256048         0.253312     0.505758     50.319494   
 std              0.139265         0.139092     0.172195     16.895269   
 min              0.010767         0.010881     0.200720     20.011738   
 25%              0.140794         0.131564     0.353609     35.655167   
 50%              0.253715         0.252823     0.514049     51.130712   
 75%              0.370674         0.373040     0.648557     64.666258   
 max              0.499989         0.49891

In [7]:
print(df.dtypes)

Theme                  object
Click Through Rate    float64
Conversion Rate       float64
Bounce Rate           float64
Scroll_Depth          float64
Age                     int64
Location               object
Session_Duration        int64
Purchases              object
Added_to_Cart          object
dtype: object


The dataset contains 1,000 records across 10 columns, with no missing values. Here's a quick summary of the numerical columns:
1. Click Through Rate: Ranges from about 0.01 to 0.50 with a mean of approximately 0.26
2. Conversion Rate: Also range from about 0.01 to 0.50 with a mean close to the Click Through Rate, approximately 0.25
3. Bounce Rate: Varies between 0.20 and 0.80, with a mean around 0.51
4. Scroll Depth: Shows a spread from 20.01 to nearly 80, with a mean of 50.32
5. Age: The age of users ranges from 18 to 65 years, with a mean age of about 41.5
6. Session Duration: This varies widely from 38 seconds to nearly 1800 seconds (30 minutes), with a mean session duration of approximately 925 seconds (about 15 minutes)

Now, let's move on to comparing the performance of both themes based on the provides metrics. We'll look into the average Click Through Rate, Conversion Rate, Bounce Rate, and other relevant metrics for each theme. Afterwards, we can perform hypothesis testing to identify if there's a statistically significant diference between the themes

In [9]:
import numpy as np

In [11]:
# Select only numeric columns for mean calculation
numeric_columns = df.select_dtypes(include=[np.number]).columns
theme_performance = df.groupby('Theme')[numeric_columns].mean()

# Sort data by conversion rate for better comparison
theme_performance_sorted = theme_performance.sort_values(by='Conversion Rate', ascending=False)

print(theme_performance_sorted)

             Click Through Rate  Conversion Rate  Bounce Rate  Scroll_Depth  \
Theme                                                                         
Light Theme            0.247109         0.255459     0.499035     50.735232   
Dark Theme             0.264501         0.251282     0.512115     49.926404   

                   Age  Session_Duration  
Theme                                     
Light Theme  41.734568        930.833333  
Dark Theme   41.332685        919.482490  


The comparison between the light Theme and Dark Theme on average performance metrics reveals the following insights
1. Click Through Rate (CTR): The Dark Theme has a slightly higher average CTR (0.2645) compared to the Light Theme (0.2471)
2. Conversion Rate: The Light leads with a marginally higher average Conversion Rate (0.2555) compared to the Dark Theme (0.2513)
3. Bounce Rate: The Bounce Rate is slightly higher for the Dark Theme (0.5121) than for the Ligth Theme (0.4990)
4. Scroll Depth: Users on the light Theme scroll slightly further on average (50.74%) compared to those on the Dark Theme (49.93%)
5. Age: The average of users is similar across themes, with the Light Theme at approximately 41.73 years and the Dark Theme at 41.33 years 
6. Session Duration: The average session duration is slightly longer for the Light Theme (930.83 seconds) than for those on the Dark Theme (919.48 seconds)

From these insights, it appears that the Light Theme slightly outperforms the Darks Theme in terms of Conversion Rate, Bounce Rate, Scroll Depth, and Session Duration, while the Dark Theme leads in click Through Rate. However, the differences are relatively minor across all metrics

## Getting Started with Hyphotesis Testing

We'll use a significance level (alpha) of 0.05 for our hypothesis testing. It means we'll consider a result statitically significant if the p-value from our test is less than 0.05

Let's start hyphotesis testing based on the Conversion Rate between the Light Theme and Dark Theme. Our hypothesis are as follows:

1. Null Hypothesis (H0): There is no difference in Conversion Rates between the Light Theme and Dark Theme
2. Alternative Hypothesis (Ha): There is a difference in Conversion Rates between the Light Theme and Dark Theme

We'll use a two-sample t-test to compare the means of the two independent samples. Let's proceed with the test

In [13]:
# extracting conversion rates for both themes

conversion_rates_light = df[df['Theme'] == 'Light Theme']['Conversion Rate']
conversion_rates_dark = df[df['Theme'] == 'Dark Theme']['Conversion Rate']

# permorming a two-sample t-test
t_stat, p_value = ttest_ind(conversion_rates_light, conversion_rates_dark, equal_var=False)

t_stat, p_value


(0.4748494462782632, 0.6349982678451778)

The result of the two-sample t-test gives a p-value of approximately 0.635. Since this p-value is much greater than our significance level of 0.05, we do not hace enough evidence to reject the null hypothesis. Therefore, we conclude that there is no statistically significant difference in Conversion Rates between the Light Theme and Dark Theme based on the data provided

Now, let's conduct hypothesis testing based on the Click Through Rate (CTR) to see if there's a statiscally significant difference between the Light Theme and Dark Theme regarding how often users click through. Our hypotheses remain structured similarly:

1. Null Hypothesis (H0): There is no difference in Click Through Rates between the Light Theme and Dark Theme
2. Altenative Hypothesis (Ha): There is a difference in Click Rates between the Light theme and Dark Theme

We'll perform a two-sample t-test on the CTR for both themes. Let's proceed with the calculation:

In [14]:
# extracting click through rates for both themes
ctr_light = df[df['Theme'] == 'Light Theme']['Click Through Rate']
ctr_dark = df[df['Theme'] == 'Dark Theme']['Click Through Rate']

# performing a two-sample t-test
t_stat_ctr, p_value_ctr = ttest_ind(ctr_light, ctr_dark, equal_var=False)

t_stat_ctr, p_value_ctr

(-1.9781708664172253, 0.04818435371010704)

The two-sample t-test for the Click Through Rate (CTR) between the Light Theme and Dark Theme yields a p-value of approximately 0.048. This p-value is slightly below our significance level of 0.05, indicating that there is a statistically significant difference in Click Throuhg Rates the Light Theme and Dark Theme, with the Dark Theme likely having a higher CTR given the direction of the statistic



Now, let's perform Hypothesis Testing Based on two other metrics: bounce rate and scroll depth, which are important metric for analyzing the performance of a theme or a desing on a website. I'll first perform these statitical tests and then create a table to show the report of all the tests we have done

In [16]:
# extracting bounce rates for both theme
bounce_rates_light = df[df['Theme'] == 'Light Theme'] ['Bounce Rate']
bounce_rates_dark = df[df['Theme'] == 'Dark Theme'] ['Bounce Rate']

# performing a two-sample t-test for bounce rate
t_stat_bounce, p_value_bounce = ttest_ind(bounce_rates_light, bounce_rates_dark, equal_var=False)

# extracting scroll depths for both themes
scroll_depth_light = df[df['Theme'] == 'Light Theme']['Scroll_Depth']
scroll_depth_dark = df[df['Theme'] == 'Dark Theme']['Scroll_Depth']

# performing a two-sample t-test for scroll depth
t_stat_scroll, p_value_scroll = ttest_ind(scroll_depth_light, scroll_depth_dark, equal_var=False)

# creating a table for comparison
comparison_table = pd.DataFrame({
    'Metric':['Click Through Rate', 'Conversion Rate', 'Bounce Rate', 'Scroll Depth'],
    'T-Statistic': [t_stat_ctr, t_stat, t_stat_bounce, t_stat_scroll],
    'P-Value': [p_value_ctr, p_value, p_value_bounce, p_value_scroll]
})

comparison_table




Metric  T-Statistic   P-Value
0  Click Through Rate    -1.978171  0.048184
1     Conversion Rate     0.474849  0.634998
2         Bounce Rate    -1.201888  0.229692
3        Scroll Depth     0.756228  0.449692

So, here's a table comparing the performance of the Light Theme and Dark Theme across various metrics based on hypothesis testing:

1. Click Through Rate: The test reveals a statistically significant difference, with the Dark Theme likely performing better (P-Value = 0.048)
2. Conversion Rate: No statistically significant difference was found (P-Value=0.635)
3. Bounce Rate: There's no statistically significant difference in Bounce Rates between the themes (P-Value=0.230)
4. Scroll Depth: Similarly, no statistically significant difference is observed in Scroll Depths (P-Value = 0.450)

In summary, while the two themes perform similarly across most metrics, the Dark Theme has a slight edge in terms of engaging users to click throuhg. For the other key performance indicators like Conversion Rate, Bounce Rate, and Scroll Depth, the choice between a Light Theme and a Dark Theme does not significatly affect user behaviour according to the data provided